gapic-common/lib/gapic/rest/paged_enumerable.rb

```diff
@@ -0,0 +1,226 @@
+    #
+    # PagedEnumerable provides the enumerations over the resource data, and also provides the enumerations over the
+    # pages themselves.
+    #
+    # @example normal iteration over resources.
+    #   paged_enumerable.each { |resource| puts resource }
+    #
+    # @example per-page iteration.
+    #   paged_enumerable.each_page { |page| puts page }
+    #
+    # @example Enumerable over pages.
+    #   paged_enumerable.each_page do |page|
+    #     page.each { |resource| puts resource }
+    #   end
+    #
+    # @example more exact operations over pages.
+    #   while some_condition()
+    #     page = paged_enumerable.page
+    #     do_something(page)
+    #     break if paged_enumerable.next_page?
+    #     paged_enumerable.next_page
+    #   end
+    #
+    class PagedEnumerable
+      include Enumerable
+
+      ##
+      # @attribute [r] page
+      #   @return [Page] The current page object.
+      attr_reader :page
+
+      ##
+      # @private
+      # @param service_stub [Object] The REST service_stub with the baseline implementation for the wrapped method.
+      # @param method_name [Symbol] The REST method name that is being wrapped.
+      # @param request [Object] The request object.
+      # @param response [Object] The response object.
+      # @param options [Gapic::CallOptions] The options for making the RPC call.
+      # @param format_resource [Proc] A Proc object to format the resource object. The Proc should accept response as an
+      #   argument, and return a formatted resource object. Optional.
+      #
+      def initialize service_stub, method_name, resource_field_name, request, response, options, format_resource: nil
+        @service_stub = service_stub
+        @method_name = method_name
+        @resource_field_name = resource_field_name
+        @request = request
+        @response = response
+        @options = options
+        @format_resource = format_resource
+
+        @page = Page.new response, resource_field_name, format_resource: @format_resource
+      end
+
+      ##
+      # Iterate over the resources.
+      #
+      # @yield [Object] Gives the resource objects in the stream.
+      #
+      # @raise [RuntimeError] if it's not started yet.
+      #
+      def each &block
+        return enum_for :each unless block_given?
+
+        each_page do |page|
+          page.each(&block)
+        end
+      end
+
+      ##
+      # Iterate over the pages.
+      #
+      # @yield [Page] Gives the pages in the stream.
+      #
+      # @raise if it's not started yet.
+      # @return [Enumerable, Nil]
+      def each_page
+        return enum_for :each_page unless block_given?
+
+        loop do
+          break if @page.nil?
+          yield @page
+          next_page!
+        end
+      end
+
+      ##
+      # True if it has the next page.
+      #
+      # @return [Boolean]
+      #
+      def next_page?
+        @page.next_page_token?
```

Comment:
`@page` gets set to `nil` once iteration is complete. Should handle that case without crashing.

Reply:
:+1: